renderdoc/driver/d3d12/d3d12_pixelhistory.cpp

```diff
@@ -246,6 +271,248 @@ void D3D12DebugManager::PixelHistoryCopyPixel(ID3D12GraphicsCommandListX *cmd,
   state.ApplyState(m_pDevice, cmd);
 }
 
+// D3D12PixelHistoryShaderCache manages temporary shaders created for pixel history.
+struct D3D12PixelHistoryShaderCache
+{
+  D3D12PixelHistoryShaderCache(WrappedID3D12Device *device, ID3DBlob *PersistentPrimIDPS,
+                               ID3DBlob *PersistentPrimIDPSDxil, ID3DBlob *FixedColorPS,
+                               ID3DBlob *FixedColorPSDxil)
+      : m_pDevice(device),
+        m_PrimIDPS(PersistentPrimIDPS),
+        m_PrimIDPSDxil(PersistentPrimIDPSDxil),
+        m_FixedColorPS(FixedColorPS),
+        m_FixedColorPSDxil(FixedColorPSDxil)
+  {
+  }
+
+  ~D3D12PixelHistoryShaderCache() {}
+
+  // Returns a fragment shader that outputs a fixed color
+  ID3DBlob *GetFixedColorShader(bool dxil) { return dxil ? m_FixedColorPSDxil : m_FixedColorPS; }
+
+  // Returns a fragment shader that outputs primitive ID
+  ID3DBlob *GetPrimitiveIdShader(bool dxil) { return dxil ? m_PrimIDPSDxil : m_PrimIDPS; }
+
+  // TODO: This class should also manage any shader replacements needed during pixel history
+
+private:
+  WrappedID3D12Device *m_pDevice;
+
+  ID3DBlob *m_PrimIDPS;
+  ID3DBlob *m_PrimIDPSDxil;
+  ID3DBlob *m_FixedColorPS;
+  ID3DBlob *m_FixedColorPSDxil;
+};
+
+// D3D12PixelHistoryCallback is a generic D3D12ActionCallback that can be used
+// for pixel history replays.
+struct D3D12PixelHistoryCallback : public D3D12ActionCallback
+{
+  D3D12PixelHistoryCallback(WrappedID3D12Device *device, D3D12PixelHistoryShaderCache *shaderCache,
+                            const D3D12PixelHistoryCallbackInfo &callbackInfo,
+                            ID3D12QueryHeap *occlusionQueryHeap)
+      : m_pDevice(device),
+        m_ShaderCache(shaderCache),
+        m_CallbackInfo(callbackInfo),
+        m_OcclusionQueryHeap(occlusionQueryHeap)
+  {
+    m_pDevice->GetQueue()->GetCommandData()->m_ActionCallback = this;
+  }
+
+  virtual ~D3D12PixelHistoryCallback()
+  {
+    m_pDevice->GetQueue()->GetCommandData()->m_ActionCallback = NULL;
+  }
+
+protected:
+  // Update the given scissor to just the pixel for which pixel history was requested.
+  void ScissorToPixel(const D3D12_VIEWPORT &view, D3D12_RECT &scissor)
+  {
+    float fx = (float)m_CallbackInfo.x;
+    float fy = (float)m_CallbackInfo.y;
+    float y_start = view.TopLeftY;
+    float y_end = view.TopLeftY + view.Height;
+    if(view.Height < 0)
```

Comment:
I don't think negative viewport height is a thing on D3D12? This seems (harmlessly) left over from the vulkan implementation.

Reply:
It was added in an Agility SDK release: https://devblogs.microsoft.com/directx/agility-sdk-1-602-0/